src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.cc

```diff
@@ -0,0 +1,604 @@
+      return extent->at_max_capacity() ?
+        split_entry(t, lo, punch_pt, extent) :
+        punch_lextent_ertr::make_ready_future<ExtMapNodeRef>(std::move(extent));
+    }).safe_then([this, &result, &t, lo, len](ExtMapNodeRef extent) mutable {
+      return extent->punch_lextent(t, lo, len).safe_then(
+        [&result](auto item_list) mutable {
+        result.splice(result.end(), item_list, item_list.begin(), item_list.end());
+      });
+    });
+  }).safe_then([result=std::move(result_up)] {
+    return punch_lextent_ret(
+           punch_lextent_ertr::ready_future_marker{},
+           std::move(*result));
+  });
+
+}
+
+ExtMapInnerNode::rm_lextent_ret
+ExtMapInnerNode::rm_lextent(Transaction &t, objaddr_t lo, lext_map_val_t val)
+{
+  auto rm_pt = get_containing_child(lo);
+  return extmap_load_extent(tm, t, rm_pt->get_val(), depth-1).safe_then(
+    [this, rm_pt, &t, lo, val=std::move(val)](auto extent) mutable {
+    if (extent->at_min_capacity()) {
+      return merge_entry(t, lo, get_containing_child(lo), extent);
+    } else {
+      return merge_entry_ertr::make_ready_future<ExtMapNodeRef>(std::move(extent));
+    }
+  }).safe_then([this, &t, lo, val](ExtMapNodeRef extent) mutable {
+    return extent->rm_lextent(t, lo, val);
+  });
+}
+
+ExtMapInnerNode::find_hole_ret
+ExtMapInnerNode::find_hole(Transaction &t, objaddr_t lo, extent_len_t len)
+{
+  logger().debug("ExtMapInnerNode::find_hole for logic_offset={}, len={}, *this={}",
+    lo, len, *this);
+  auto bounds = bound(lo, lo + len);
+  return seastar::do_with(
+    bounds.first,
+    bounds.second,
+    lext_map_val_t{L_ADDR_NULL, 0, 0},
+    [this, &t, lo, len](auto &i, auto &e, auto &ret) {
+      return crimson::do_until(
+	[this, &t, &i, &e, &ret, lo, len] {
+	  if (i == e + 1) {
+	    return find_hole_ertr::make_ready_future<ExtentRef>(
+			    std::make_unique<Extent>(lo, L_ADDR_NULL, 0, 0));
+	  }
+	  return extmap_load_extent(tm, t, i->get_val(), depth-1)
+	    .safe_then([this, &t, &i, lo, len](auto extent) mutable {
+	    logger().debug(
+	      "ExtMapInnerNode::find_hole in {} for lo {} len {} ",
+	      *extent, lo, len);
+	    return extent->find_hole(t, lo, len);
+	  }).safe_then([lo, &i, &ret](auto ext_map) mutable {
+	    i++;
+	    if (ext_map->laddr != L_ADDR_NULL) {
+	      ret.laddr = ext_map->laddr;
+	      ret.lextent_offset = ext_map->lextent_offset;
+	      ret.length = ext_map->length;
+	    }
+	    return find_hole_ertr::make_ready_future<ExtentRef>(
+	           std::make_unique<Extent>(lo, ret.laddr,
+				   ret.lextent_offset, ret.length));
+	  });
+        }).safe_then([lo, &ret]() {
+          return find_hole_ertr::make_ready_future<ExtentRef>(
+                 std::make_unique<Extent>(lo, ret.laddr,
+                                 ret.lextent_offset, ret.length));
+	});
+    });
+}
+
+ExtMapInnerNode::split_children_ret
+ExtMapInnerNode::make_split_children(Transaction &t)
+{
```

Comment:
Probably worth creating a helper to allocate N blocks rather than use nested continuations like this.